[quant] Run weight_post_process for QAT

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34232 [quant] Fix histogram observer to work with QAT on GPU
* **#33852 [quant] Run weight_post_process for QAT**

Summary:
This fixes an issue for QAT models. During eval if we call `prepare_qat` and `convert` before calling `load_state_dict` it throws an error because the weight info (num channels) is not updated in the observer module.
It is not an issue for per-tensor case

Fixes issue #33830

Test Plan:
python test/test_quantization.py EagerModePostTrainingQuantTest.test_eval_after_train
python test/test_quantization.py EagerModeQuantizationAwareTrainingTest.test_eval_after_train

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20212996](https://our.internmc.facebook.com/intern/diff/D20212996)

supriyar committed:
[quant] Run weight_post_process for QAT

Summary:
This fixes an issue for QAT models. During eval if we call `prepare_qat` and `convert` before calling `load_state_dict` it throws an error because the weight info (num channels) is not updated in the observer module.
It is not an issue for per-tensor case

Fixes issue #33830

Test Plan:
python test/test_quantization.py EagerModePostTrainingQuantTest.test_eval_after_train
python test/test_quantization.py EagerModeQuantizationAwareTrainingTest.test_eval_after_train

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -232,6 +232,7 @@ def from_float(cls, mod):
             assert hasattr(mod, 'activation_post_process'), \
                 'Input QAT module must have observer attached'
             weight_post_process = mod.weight_fake_quant
+            weight_post_process(mod.weight)
```
can you just move this outside of if?

jerryzh168 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -204,6 +204,7 @@ def from_float(cls, mod):
         if hasattr(mod, 'weight_fake_quant'):
             # assert type(mod) == QATLinear, 'training mode nnq.Linear.from_float only works for nn.qat.Linear'
             weight_post_process = mod.weight_fake_quant
+            weight_post_process(mod.weight)
```
similar for this one

jerryzh168 commented on test/test_quantization.py:
```diff
@@ -793,6 +813,31 @@ def checkQuantized(model):
         model = quantize_qat(model, test_only_train_fn, self.img_data)
         checkQuantized(model)
 
+    def test_eval_after_train(self):
```
I think maybe include all steps would be better: e.g. test_train_save_load_eval

supriyar committed:
Update on "[quant] Run weight_post_process for QAT"


Summary:
This fixes an issue for QAT models. During eval if we call `prepare_qat` and `convert` before calling `load_state_dict` it throws an error because the weight info (num channels) is not updated in the observer module.
It is not an issue for per-tensor case

Fixes issue #33830

Test Plan:
python test/test_quantization.py EagerModePostTrainingQuantTest.test_eval_after_train
python test/test_quantization.py EagerModeQuantizationAwareTrainingTest.test_eval_after_train

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -793,6 +813,31 @@ def checkQuantized(model):
+        test_only_train_fn(model, self.train_data)
+        model = convert(model)
+
+        model = quantize_qat(ManualLinearQATModel(), test_only_train_fn,
```
Looks like lines 818-822 are redundant as you are creating a model afresh here.

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -793,6 +813,31 @@ def checkQuantized(model):
         checkQuantized(model)
 
+    def test_train_save_load_eval(self):
+        qengine = 'fbgemm'
```
Can you sweep over both qengines here?

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -793,6 +813,31 @@ def checkQuantized(model):
         model = quantize_qat(model, test_only_train_fn, self.img_data)
         checkQuantized(model)
 
+    def test_train_save_load_eval(self):
```
Should this function be called test_train_save_load_qat(self):?

supriyar replied:
The class has information about PTQ vs QAT  - `EagerModeQuantizationAwareTrainingTest`

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -311,6 +311,26 @@ def checkQuantized(model):
         checkQuantized(model)
 
+    def test_train_save_load_eval(self):
+        model = AnnotatedTwoLayerLinearModel()
```
Suggest sweeping over multiple qconfigs here (cover both per tensor and per channel quant)

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -311,6 +311,26 @@ def checkQuantized(model):
 
         checkQuantized(model)
 
+    def test_train_save_load_eval(self):
```
Should this be called test_post_train_quant_save_load(self):?

supriyar replied:
The class has information about PTQ vs QAT - `EagerModePostTrainingQuantTest`

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -793,6 +813,31 @@ def checkQuantized(model):
+        model.load_state_dict(quant_state_dict)
+
+        out = model(x)
+        self.assertEqual(ref, out)
```
How about testing for consistency at multiple points: 
1. Consistency of fake-quant state dict with loading a model from scratch
2. Consistency of state dict after convert with a model created using PTQ (the qconfig needs to be consistent: Both FQ and observer need to use the same type of quantization).

raghuramank100 reviewed (changes requested):
Good find! I have a couple of suggestions to improve the tests. Thanks

supriyar committed:
Update on "[quant] Run weight_post_process for QAT"


Summary:
This fixes an issue for QAT models. During eval if we call `prepare_qat` and `convert` before calling `load_state_dict` it throws an error because the weight info (num channels) is not updated in the observer module.
It is not an issue for per-tensor case

Fixes issue #33830

Test Plan:
python test/test_quantization.py EagerModePostTrainingQuantTest.test_eval_after_train
python test/test_quantization.py EagerModeQuantizationAwareTrainingTest.test_eval_after_train

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -311,6 +311,41 @@ def checkQuantized(model):
 
         checkQuantized(model)
 
+    @given(qengine=st.sampled_from(("qnnpack", "fbgemm")))
```
NIT: Can we have a better name than train_save_load_eval? There is no training here.

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -793,6 +828,43 @@ def checkQuantized(model):
+            fq_state_dict = model.state_dict()
+
+            test_only_train_fn(model, self.train_data)
+            model = convert(model)
```
Dont you need to call model.eval() here?

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -793,6 +828,43 @@ def checkQuantized(model):
+            # Check to make sure the model after prepare_qat has the same state_dict as original.
+            self.assertEqual(set(fq_state_dict.keys()), set(new_state_dict.keys()))
+
+            torch.quantization.convert(model, inplace=True)
```
Similar comment: you need to call model.eval() here

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -793,6 +828,43 @@ def checkQuantized(model):
+            self.assertEqual(set(fq_state_dict.keys()), set(new_state_dict.keys()))
+
+            torch.quantization.convert(model, inplace=True)
+            model.load_state_dict(quant_state_dict)
```
How about adding one more check here?
```python
            model = TwoLayerLinearModel()
            model.eval()
            model = torch.quantization.QuantWrapper(model)
            model.qconfig = torch.quantization.get_default_qconfig(qengine)
            torch.quantization.prepare(model, inplace=True)
            torch.quantization.convert(model, inplace=True)
            self.assertEqual(set(model.state_dict().keys()), set(quant_state_dict.keys()))
            out = model(x)
            self.assertEqual(ref, out)
```

raghuramank100 reviewed (changes requested):
Looks good, adding a request for additional testing.

supriyar committed:
Update on "[quant] Run weight_post_process for QAT"


Summary:
This fixes an issue for QAT models. During eval if we call `prepare_qat` and `convert` before calling `load_state_dict` it throws an error because the weight info (num channels) is not updated in the observer module.
It is not an issue for per-tensor case

Fixes issue #33830

Test Plan:
python test/test_quantization.py EagerModePostTrainingQuantTest.test_eval_after_train
python test/test_quantization.py EagerModeQuantizationAwareTrainingTest.test_eval_after_train

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Run weight_post_process for QAT"


Summary:
This fixes an issue for QAT models. During eval if we call `prepare_qat` and `convert` before calling `load_state_dict` it throws an error because the weight info (num channels) is not updated in the observer module.
It is not an issue for per-tensor case

Fixes issue #33830

Test Plan:
python test/test_quantization.py EagerModePostTrainingQuantTest.test_eval_after_train
python test/test_quantization.py EagerModeQuantizationAwareTrainingTest.test_eval_after_train

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 reviewed (approved):
Looks good, thanks for making the changes!

supriyar committed:
Update on "[quant] Run weight_post_process for QAT"


Summary:
This fixes an issue for QAT models. During eval if we call `prepare_qat` and `convert` before calling `load_state_dict` it throws an error because the weight info (num channels) is not updated in the observer module.
It is not an issue for per-tensor case

Fixes issue #33830

Test Plan:
python test/test_quantization.py EagerModePostTrainingQuantTest.test_eval_after_train
python test/test_quantization.py EagerModeQuantizationAwareTrainingTest.test_eval_after_train

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Run weight_post_process for QAT"


Summary:
This fixes an issue for QAT models. During eval if we call `prepare_qat` and `convert` before calling `load_state_dict` it throws an error because the weight info (num channels) is not updated in the observer module.
It is not an issue for per-tensor case

Fixes issue #33830

Test Plan:
python test/test_quantization.py EagerModePostTrainingQuantTest.test_eval_after_train
python test/test_quantization.py EagerModeQuantizationAwareTrainingTest.test_eval_after_train

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20212996](https://our.internmc.facebook.com/intern/diff/D20212996)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Run weight_post_process for QAT"


Summary:
This fixes an issue for QAT models. During eval if we call `prepare_qat` and `convert` before calling `load_state_dict` it throws an error because the weight info (num channels) is not updated in the observer module.
It is not an issue for per-tensor case

Fixes issue #33830

Test Plan:
python test/test_quantization.py EagerModePostTrainingQuantTest.test_eval_after_train
python test/test_quantization.py EagerModeQuantizationAwareTrainingTest.test_eval_after_train

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20212996](https://our.internmc.facebook.com/intern/diff/D20212996)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Run weight_post_process for QAT"


Summary:
This fixes an issue for QAT models. During eval if we call `prepare_qat` and `convert` before calling `load_state_dict` it throws an error because the weight info (num channels) is not updated in the observer module.
It is not an issue for per-tensor case

Fixes issue #33830

Test Plan:
python test/test_quantization.py EagerModePostTrainingQuantTest.test_eval_after_train
python test/test_quantization.py EagerModeQuantizationAwareTrainingTest.test_eval_after_train

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20212996](https://our.internmc.facebook.com/intern/diff/D20212996)

[ghstack-poisoned]
